### PR TITLE
Feature/147571 distributed dsa node

### PR DESCRIPTION
**Pull request by dmitriyisaev**

### CODE REVIEW

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by c9alex** [comment]
Failed tests: http://jenkins.acuitynext.net/job/Branches-DistributedData/

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

File: app/models/akka/RootNodeActor.scala [hunk 1]
```diff
@@ -124,14 +124,14 @@ object RootNodeActor {
   /**
    * Creates a new instance of [[RootNodeActor]] props.
    */
-  def props = Props(new RootNodeActor)
+  def props() = Props(new RootNodeActor())
```
Comment (uralian) [inline app/models/akka/RootNodeActor.scala:128]: just curious - why? :)

File: app/models/akka/StandardActions.scala [hunk 2]
```diff
@@ -1,40 +1,54 @@
...
+import models.api.{DSAAction, DSANode, DSAValueType}
+import models.rpc.DSAValue
 
 /**
  * Standard node actions.
  */
 object StandardActions {
   import DSAValueType._
 
+  case class ActionDescription(name:String, displayName:String, action:DSAAction)
```
Comment (uralian) [inline app/models/akka/StandardActions.scala:14]: good idea, much better that way

**Review by c9alex — DISMISSED** [review]
Failed tests:
http://jenkins.acuitynext.net/job/Branches-DistributedData/

File: app/models/api/DistributedNodesRegistry.scala [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+import com.google.inject.Inject
+import models.Settings.QueryTimeout
+import models.akka.StandardActions
+import models.rpc.DSAValue.StringValue
+import models.rpc.{InvokeRequest, RemoveRequest, SetRequest}
+
+import scala.annotation.tailrec
+import scala.concurrent.Future
+
+class DistributedNodesRegistry @Inject()(val replicator: ActorRef)(implicit cluster: Cluster, system: ActorSystem) extends Actor with ActorLogging {
```
Comment (uralian) [inline app/models/api/DistributedNodesRegistry.scala]: Do we need `cluster` as an implicit parameter? We can just take it from the `system`...

File: app/models/api/DistributedNodesRegistry.scala [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+
+  /**
+    * create node actor for path in parent / root context
+    * if parent is not created - creates it
+    *
+    * @param p path
+    * @return created DSAnode actor reg
+    */
+  private[this] def getOrCreateNode(p: String): DSANode = {
```
Comment (uralian) [inline app/models/api/DistributedNodesRegistry.scala]: perhaps move this inside the registry itself? not critical though...

**Review by uralian — APPROVED** [review]

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

File: app/models/akka/RootNodeActor.scala (hunk 1 above)
Comment (dmitriyisaev) [reply app/models/akka/RootNodeActor.scala:128]: Idea code style checker asked me))) As far as i remember

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Do not merge!!

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Comment by ilya-dubovskiy** [comment]
Build and tests are GOOD

**Pull request merged**